x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/ingest/SetSecurityUserProcessor.java

```diff
@@ -134,7 +141,8 @@ public SetSecurityUserProcessor create(Map<String, Processor.Factory> processorF
         FULL_NAME,
         EMAIL,
         ROLES,
-        METADATA;
+        METADATA,
+        API_KEY_NAME;
```

Comment:
We should take this opportunity to add other information to the processor at the same time.

I'm thinking maybe:

- API Key Id
- API Key Realm
- Authentication Realm (Name + Type)
- Authentication Type (`AuthenticationType`)

However, I suspect it's not worth having a separate property for _every_ field, may just
- `API_KEY` (and we always populate the id, name and realm) 
- `REALM` (which has realm name & type)
- `AUTHENTICATION_TYPE` (a single field)

we _could_ merge realm and authc-type, but it's probably not worth it.

Reply:
Correct me if I am wrong, if we use existing `Authentication` object to populate these fields for api key authentication , some of the fields will contain the same hard-coded information, i.e.:
```
"api_key": {"id": ..., "name": ..., "realm": "_es_api_key"},
"realm": {"name": "_es_api_key", "type": "_es_api_key"},
"authentication_type": "API_KEY"
```
For other authentication mechanisms, the `api_key` will not exist and other fields will be populated accordingly.

Reply:
I think we are risking to expose too much internals, in the hope they might be useful, and then we'll get stuck with having to treat changes to them as breaking. The specific thing I have in mind is the `_es_api_key`keyword and how it fills realm type fields to mark certain internal conditions.

Starting from what Tim suggested, I think we should add the **realm name and type as a single object**. This should work for API keys as well as plain user authn (@ywangd see `ApiKeyService#getCreatorRealmName` which only works for the name not for the type). Separately, I would suggest we add the **api key name and id as a single object**.

I don't feel strongly about having objects instead of top level fields, either way is fine to me (i.e. `realm_name` and `realm_type` as fields are OK). But I would avoid showing the `realm` parameter twice (under api keys and user), or exposing the `_es_api_key` value.

Reply:
Also, we should take into consideration Elastic Common Schema (ECS - here is a good starting point https://www.elastic.co/blog/introducing-the-elastic-common-schema) when defining new fields.

What would people say about renaming existing fields, to better align to the ECS specification?

Reply:
@albertzaharovits By "single object", do you mean the form of`"api_key": {"id": ..., "name": ...}`? 

I agree that exposing the value of `_es_api_key` will probably do more harm than good. This is the reason why I was trying to confirm in the original comment. So we could change the information object to something like the follows:

```json
"api_key": {"id": ..., "name": ..., "realm": "_es_api_key"},
"realm": {"name": "default_native", "type": "native"},
"authentication_type": "API_KEY"
```

Note that the realm object refers to the **creator** realm of the API key instead of the API key realm itself. In this case, we pretend the api key is generated by authenticating against the native realm. 

Just some more discussion point: I am aware it is techically incorrect to say "API key realm", since there is actually no such realm. But I wonder whether it would be feasible and more consistent to implement both token and API key auth as true realms? This could be a separate discussion and there is no need to tie it to this issue.

Reply:
> By "single object", do you mean the form of"api_key": {"id": ..., "name": ...}?

Yes, as opposed to `"api_key_id": ..., "api_key_name": ...` .

From the previous comment of yours I don't follow why `_es_api_key` must be present in `"api_key": {"id": ..., "name": ..., "realm": "_es_api_key"}`.

> Note that the realm object refers to the creator realm of the API key instead of the API key realm itself.

I think this is OK. API keys "impersonate" existing users, although they might have narrower privileges. In this case, the `realm` as set by the `Set Security User Ingest Processor`  is the "realm" of the creator user, which is the user the API key is impersonating. Also, we don't have a good way of expressing that the API key has the exact or less permissions of the creator user.

> But I wonder whether it would be feasible and more consistent to implement both token and API key auth as true realms?

I don't think so. Tokens, especially, are 100% designed to impersonate users and authenticating with a token or the user's credentials should be looking identical. An argument can be made that API **keys** are not part of a **user realm**, but API keys inherit the permissions of the creator so the "realm" of an API key can be defined as the realm of the creating user.

I think this conv might benefit from high bandwidth conv over Zoom. Maybe next team meeting?

Reply:
> From the previous comment of yours I don't follow why `_es_api_key` must be present in `"api_key": {"id": ..., "name": ..., "realm": "_es_api_key"}`.
> 
It does not have to present, was just a proposal.

Reply:
> Also, we should take into consideration Elastic Common Schema (ECS - here is a good starting point https://www.elastic.co/blog/introducing-the-elastic-common-schema) when defining new fields.
> 
> What would people say about renaming existing fields, to better align to the ECS specification?

This is an interesting point. Thanks for pointing this out. The current `user` object is mostly align to ECS, except we have `user.username` instead of `user.name`. Also we have an extra `user.metadata` field which can be considered as custom fields.

The new api key info will also be custom fields. We are conforming to the schema by using lower case, underscore concatenated keys. And the contents cannot be mapped to any existing core or extended fields. There is `user.domain` from ECS which is meant for LDAP or AD domain name. It is kinda like `realm` except it is a keyword type while realm is a object type. 

Overall I think what we have proposed so far is conforming to the ECS.